internal/abci/abci.go

```diff
@@ -700,12 +730,16 @@ func (a *AbciApp) PrepareProposal(ctx context.Context, req *abciTypes.RequestPre
 		zap.Int64("height", req.Height),
 		zap.Int("txs", len(req.Txs)))
 
-	addr, err := a.getValidatorPubKeyByAddr(ctx, proposerAddrToString(req.ProposerAddress), &req.Height)
-	if err != nil {
-		return nil, fmt.Errorf("failed to find proposer pubkey corresponding to %s", proposerAddrToString(req.ProposerAddress))
+	pubKey, ok := a.validatorAddressToPubKey[proposerAddrToString(req.ProposerAddress)]
+	if !ok {
+		// there is an edge case where cometbft will allow a node to PrepareProposal
+		// even if it is not a validator, if it was a validator in the most recent block
+		// we check for this here and do not propose if the node is not a validator
+		// we will also need to check for this in ProcessProposal, in case of byzantine behavior
+		return nil, fmt.Errorf("unknown proposer: failed to find proposer pubkey corresponding to address %v", req.ProposerAddress)
```

Comment:
Alright, I suppose, but since this is `PrepareProposal`, the `ProposerAddress` is *us*, the local validator, so we could surely retrieve the pubkey somehow.

Reply:
Right. I'll change the error message, but you would agree this is the correct logic, correct?